### PR TITLE
Astrid sync

### PR DESCRIPTION
**Pull request by battlmonstr**

### CODE REVIEW

File: consensus/bor/heimdallgrpc/milestone.go [hunk 1]
```diff
@@ -24,9 +24,10 @@ func (h *HeimdallGRPCClient) FetchMilestoneCount(ctx context.Context) (int64, er
 	return res.Result.Count, nil
 }
 
-func (h *HeimdallGRPCClient) FetchMilestone(ctx context.Context) (*milestone.Milestone, error) {
+func (h *HeimdallGRPCClient) FetchMilestone(ctx context.Context, number int64) (*milestone.Milestone, error) {
 	h.logger.Info("Fetching milestone")
 
+	// TODO: use number
```
Comment (taratorio) [inline consensus/bor/heimdallgrpc/milestone.go:30]: do we need to sort this one out before merging into devel?
depends on whether http or graph is used when running heimdall and erigon on the same process

Comment (taratorio) [reply consensus/bor/heimdallgrpc/milestone.go:30]: in `etc/backend.go` we have this which means we can't merge in devel until the grpc is changed?

```
        var heimdallClient heimdall.IHeimdallClient
	if chainConfig.Bor != nil {
		if !config.WithoutHeimdall {
			if config.HeimdallgRPCAddress != "" {
				heimdallClient = heimdallgrpc.NewHeimdallGRPCClient(config.HeimdallgRPCAddress, logger)
			} else {
				heimdallClient = heimdall.NewHeimdallClient(config.HeimdallURL, logger)
			}
		}

		flags.Milestone = config.WithHeimdallMilestones
	}
```

File: polygon/sync/heimdall_test.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	"github.com/ledgerwatch/erigon/consensus/bor/heimdall/checkpoint"
+	"github.com/ledgerwatch/erigon/consensus/bor/heimdall/milestone"
+	heimdallmock "github.com/ledgerwatch/erigon/consensus/bor/heimdall/mock"
+	"github.com/ledgerwatch/log/v3"
+	"github.com/stretchr/testify/assert"
+	"math/big"
+	"testing"
+	"time"
+)
+import "github.com/stretchr/testify/require"
```
Comment (taratorio) [inline polygon/sync/heimdall_test.go:16]: can move this in the import group above

File: consensus/bor/heimdallgrpc/milestone.go (hunk 1 above)
Comment (taratorio) [reply consensus/bor/heimdallgrpc/milestone.go:30]: Actually just noticed the change is backward compatible so only the new Astrid package won't work with the grpc which is not a blocker to merge this. We would need to update those protos and implementations at some point soonish though.

**Review by taratorio — APPROVED** [review]

File: consensus/bor/heimdallgrpc/milestone.go (hunk 1 above)
Comment (battlmonstr) [reply consensus/bor/heimdallgrpc/milestone.go:30]: yes, I've added to my TODO list for later

File: polygon/sync/heimdall_test.go (hunk 2 above)
Comment (battlmonstr) [reply polygon/sync/heimdall_test.go:16]: yes, I'll fix in my next PR, not sure why GoLand did it like so, I'll adjust the settings

**Pull request merged**